### PR TITLE
Respect member-order="bysource" even when no analyzer is present

### PR DESCRIPTION
**Pull request by eric-wieser**
This takes advantage of the fact that in Python 3.6 and up, the __dict__ attribute of classes and modules is insertion-ordered, which is usually a good approximation of "source order".

Notably, this means that `..automodule:: some_c_extension_module` will document members in the order they were added to the module dict, which is likely what the user wants.

Motivated by the `member-order` being ignored at https://docs.cocotb.org/en/latest/library_reference.html#the-cocotb-simulator-module

### Feature or Bugfix

- Feature

Help would be appreciated testing this (once I have it passing existing tests) - I don't know how to force an analyzer to be absent other than via a C extension module, and I don't really want to have sphinx create one of those just for testing!

### CODE REVIEW

File: sphinx/ext/autodoc/__init__.py [hunk 1]
```diff
@@ -693,17 +693,26 @@ def document_members(self, all_members: bool = False) -> None:
...
+                def keyfunc(entry: Tuple[Documenter, bool]) -> int:
+                    fullname = entry[0].name.split('::')[1]
+                    return tagorder.get(fullname, len(tagorder))
+                memberdocumenters.sort(key=keyfunc)
+            else:
+                # Assume that member discovery order matches source order.
+                # This is a reasonable assumption in Python 3.6 and up, where
+                # module.__dict__ is insertion-ordered.
+                pass
+        elif member_order == 'alphabetic':
```
Comment (tk0miya) [inline sphinx/ext/autodoc/__init__.py]: I noticed our doc says the correct value is `'alphabetical'`. And I feel it is better wording.
https://www.sphinx-doc.org/en/master/usage/extensions/autodoc.html#confval-autodoc_member_order

File: sphinx/ext/autodoc/__init__.py [hunk 2]
```diff
@@ -693,17 +693,26 @@ def document_members(self, all_members: bool = False) -> None:
...
+            memberdocumenters.sort(key=lambda e: e[0].name)
+        else:
+            raise ValueError("Illegal member order {}".format(member_order))
```
Comment (tk0miya) [inline sphinx/ext/autodoc/__init__.py:721]: I'll post another PR to warn such wrong options. So you don't need to handle it here.

File: sphinx/ext/autodoc/__init__.py [hunk 3]
```diff
@@ -693,17 +693,26 @@ def document_members(self, all_members: bool = False) -> None:
...
+
+                def keyfunc(entry: Tuple[Documenter, bool]) -> int:
+                    fullname = entry[0].name.split('::')[1]
+                    return tagorder.get(fullname, len(tagorder))
+                memberdocumenters.sort(key=keyfunc)
+            else:
+                # Assume that member discovery order matches source order.
+                # This is a reasonable assumption in Python 3.6 and up, where
+                # module.__dict__ is insertion-ordered.
+                pass
```
Comment (tk0miya) [inline sphinx/ext/autodoc/__init__.py:717]: We need to consider the order of the members not included in `__dict__`.

* enum members (they are not in `__dict__`)
* `__slots__`
* `__annotations__` (type annotated variables, but not initialized yet)
* instance variables
* members defined at ancestor classes
* `module.__all__` (which is strong `__dict__` or `__all__`?)

Note: Now we don't use `__dict__` for discovery. So we have to order them manually if we adopt `__dict__`.

**Review by tk0miya — COMMENTED** [review]
Thank you for proposal. This is my headache task for next release. Good for discussing.

File: sphinx/ext/autodoc/__init__.py (hunk 1 above)
Comment (eric-wieser) [reply sphinx/ext/autodoc/__init__.py]: That may be, but all the tests use this spelling

File: sphinx/ext/autodoc/__init__.py (hunk 3 above)
Comment (eric-wieser) [reply sphinx/ext/autodoc/__init__.py:717]: Bear in mind this is just a backup option for when no analyzer is available. You're right that this is not a perfect ordering, but sorting alphabetically when "by source" is requested is far worse than a rough approximation to by source.

---

> enum members (they are not in __dict__)

I disagree:
```
>>> C = enum.Enum("Color", "red green blue")
>>> C.__dict__\
mappingproxy({'_generate_next_value_': <function enum.Enum._generate_next_value_(name, start, count, last_values)>,
              '__doc__': 'An enumeration.',
              '__module__': '__main__',
              '_member_names_': ['red', 'green', 'blue'],
              '_member_map_': {'red': <Color.red: 1>,
               'green': <Color.green: 2>,
               'blue': <Color.blue: 3>},
              '_member_type_': object,
              '_value2member_map_': {1: <Color.red: 1>,
               2: <Color.green: 2>,
               3: <Color.blue: 3>},
              'red': <Color.red: 1>,
              'green': <Color.green: 2>,
              'blue': <Color.blue: 3>,
              '__new__': <function enum.Enum.__new__(cls, value)>})
```

> `__slots__`

Also not true:
```
>>> class Foo:
...     __slots__ = ('a', 'b')

>>> Foo.__dict__
mappingproxy({'__module__': '__main__',
              '__slots__': ('a', 'b'),
              'a': <member 'a' of 'Foo' objects>,
              'b': <member 'b' of 'Foo' objects>,
              '__doc__': None})
```

> `__annotations__` (type annotated variables, but not initialized yet)

True, but likely irrelevant for C extension modules

> instance variables

True

> members defined at ancestor classes

True

> `module.__all__`

Likely irrelevant for C extension modules

File: sphinx/ext/autodoc/__init__.py (hunk 2 above)
Comment (eric-wieser) [reply sphinx/ext/autodoc/__init__.py:721]: Will that PR convert wrong options into right ones?

File: sphinx/ext/autodoc/__init__.py (hunk 1 above)
Comment (tk0miya) [reply sphinx/ext/autodoc/__init__.py]: I posted https://github.com/sphinx-doc/sphinx/pull/7679 to migrate the old name to new one with a warning. I suppose it helps to migrate to new options.

File: sphinx/ext/autodoc/__init__.py (hunk 2 above)
Comment (tk0miya) [reply sphinx/ext/autodoc/__init__.py:721]: No, it only emits a warning.

**Comment by eric-wieser** [comment]
Can you reopen this?

**Comment by tk0miya** [comment]
sorry. it's mistake.

File: sphinx/ext/autodoc/__init__.py (hunk 2 above)
Comment (eric-wieser) [reply sphinx/ext/autodoc/__init__.py:721]: So what should I do if `member_order == "nonsense"` here?

Comment (tk0miya) [reply sphinx/ext/autodoc/__init__.py:721]: Now it is treated as "alphabetical" (default). I think no reason to change it from now. But I'm also okay if we change the behavior to "not sorted".

If we choose to raise an error, it is better to do that in more earlier stage (ex. config-inited).

File: sphinx/ext/autodoc/__init__.py (hunk 3 above)
Comment (tk0miya) [reply sphinx/ext/autodoc/__init__.py:717]: >enum members (they are not in dict)

Sorry, this is my wrong. I mistook `dir()` for `__dict__`. The result of `dir()` for enum classes does not contain methods. But it is not related to sorting. Please forget.
```
Python 3.8.3 (default, May 15 2020, 11:21:44)
[Clang 11.0.3 (clang-1103.0.32.59)] on darwin
Type "help", "copyright", "credits" or "license" for more information.
>>> import enum
>>> class MyEnum(enum.Enum):
...     foo = 1
...     bar = 2
...     def hello(self): pass
...
>>> dir(MyEnum)
['__class__', '__doc__', '__members__', '__module__', 'bar', 'foo']
```

> __slots__

This is also my wrong.

Comment (tk0miya) [reply sphinx/ext/autodoc/__init__.py:717]: Note: We have some issues for sorting members: https://github.com/sphinx-doc/sphinx/issues/5270, https://github.com/sphinx-doc/sphinx/issues/3673, https://github.com/sphinx-doc/sphinx/pull/7633, https://github.com/sphinx-doc/sphinx/pull/7177 and so on.
So I'm thinking about the whole of sorting system for autodoc.

But they are not a blocker of this improvement. So no reason to merge this. I'll review this again later (maybe tonight).

File: sphinx/ext/autodoc/__init__.py (hunk 1 above)
Comment (eric-wieser) [reply sphinx/ext/autodoc/__init__.py]:
```suggestion
        elif member_order == 'alphabetical':
```

File: sphinx/ext/autodoc/__init__.py (hunk 3 above)
Comment (eric-wieser) [reply sphinx/ext/autodoc/__init__.py:717]: Thanks for those links, I agree that this is not an end-all solution, but I think removing the initial alphabetic sort is probably a good first step.

I've rebased this PR after your `alphabetical` spelling fix.

**Comment by tk0miya** [comment]
Merged. Thank you always!

**Pull request merged**